vyper/semantics/types/utils.py

```diff
@@ -82,21 +85,28 @@ def _failwith(type_name):
             f"No builtin or user-defined type named '{type_name}'. {suggestions_str}", node
         ) from None
 
+    def _check_location(typ):
+        if location in typ._invalid_locations:
+            raise InvalidType(f"{typ} is not instantiable in {location.name.lower()}", node)
+
     if isinstance(node, vy_ast.Tuple):
         tuple_t = namespace["$TupleT"]
-
+        _check_location(tuple_t)
```

Comment:
i think this should either be refactored to a `finalize()` helper function or even better - `type_from_annotation` calls a `_type_from_annotation` helper function and then checks the result.